pkg/accounting/accounting.go

```diff
@@ -0,0 +1,212 @@
+// license that can be found in the LICENSE file.
+
+package accounting
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/p2p"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+var _ Interface = (*Accounting)(nil)
+
+// Interface is the main interface for Accounting
+type Interface interface {
+	// Reserve reserves a portion of the balance for peer
+	// It returns an error if the operation would risk exceeding the disconnect threshold
+	// This should be called (always in combination with Release) before a Credit action to prevent overspending in case of concurrent requests
+	Reserve(peer swarm.Address, price uint64) error
+	// Release releases reserved funds
+	Release(peer swarm.Address, price uint64)
+	// Credit increases the balance the peer has with us (we "pay" the peer)
+	Credit(peer swarm.Address, price uint64) error
+	// Debit increases the balance we have with the peer (we get "paid")
+	Debit(peer swarm.Address, price uint64) error
+	// Balance returns the current balance for the given peer
+	Balance(peer swarm.Address) (int64, error)
+}
+
+// PeerBalance holds all relevant accounting information for one peer
+type PeerBalance struct {
+	lock     sync.Mutex
+	balance  int64  // amount that the peer owes us if positive, our debt if negative
+	reserved uint64 // amount currently reserved for active peer interaction
+}
+
+// Options for accounting
+type Options struct {
+	DisconnectThreshold uint64
+	Logger              logging.Logger
+	Store               storage.StateStorer
+}
+
+// Accounting is the main implementation of the accounting interface
+type Accounting struct {
+	balancesMu          sync.Mutex // mutex for accessing the balances map
+	balances            map[string]*PeerBalance
+	logger              logging.Logger
+	store               storage.StateStorer
+	disconnectThreshold uint64 // the debt threshold at which we will disconnect from a peer
+}
+
+// NewAccounting creates a new Accounting instance with the provided options
+func NewAccounting(o Options) *Accounting {
+	return &Accounting{
+		balances:            make(map[string]*PeerBalance),
+		disconnectThreshold: o.DisconnectThreshold,
+		logger:              o.Logger,
+		store:               o.Store,
+	}
+}
+
+// Reserve reserves a portion of the balance for peer
+func (a *Accounting) Reserve(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	// the previously reserved balance plus the new price is the maximum amount paid if all current operations are successful
+	// since we pay this we have to reduce this (positive quantity) from the balance
+	// the disconnectThreshold is stored as a positive value which is why it must be negated prior to comparison
+	if balance.balance-int64(balance.reserved+price) < -int64(a.disconnectThreshold) {
```

Comment:
This is not the correct check.
It would have to be `-int64(a.disconnectThreshold) > balance.netBalance()-int64(price)` which is not much clearer either.

Reply:
I added `netBalance` as `freeBalance` which makes it at least somewhat more legible.